### PR TITLE
HTTPCLIENT-2277: Revision and simplification of the cache eviction logic

### PR DESCRIPTION
**Pull request by ok2c**
This change-set revises and simplified the cache eviction logic to ensure conformance with RFC 9111 section 4.4 

@arturobernalg Please review.

### CODE REVIEW

**Comment by arturobernalg** [comment]
> This change-set revises and simplified the cache eviction logic to ensure conformance with RFC 9111 section 4.4
> 
> @arturobernalg Please review.

Sure, just I’ll need time

**Review by arturobernalg — APPROVED** [review]
@ok2c  The change-set looks good. Just keep in mind my comment.

File: httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/BasicHttpCache.java [hunk 1]
```diff
@@ -313,36 +324,64 @@ public CacheHit storeReusing(
...
         }
-        try {
-            storage.removeEntry(rootKey);
-        } catch (final ResourceIOException ex) {
-            if (LOG.isWarnEnabled()) {
-                LOG.warn("I/O error removing cache entry with key {}", rootKey);
+        removeInternal(rootKey);
+        if (root.isVariantRoot()) {
+            for (final String variantKey : root.getVariantMap().values()) {
+                LOG.debug("Evicting variant cache entry {}", variantKey);
```
[the remark this answers is not in the document]
Comment (stanio) [reply httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/BasicHttpCache.java:334]: > Consider wrapping all debug log statements with a ´if (LOG.isDebugEnabled())´ check for performance enhancement by avoiding unnecessary string concatenation when debug logging is not enabled.

There's no string concatenation/formatting happening with parametrized messages when the target log level is not enabled:

-   [What is the fastest way of (not) logging?](https://slf4j.org/faq.html#logging_performance)

[the remark this answers is not in the document]
Comment (arturobernalg) [reply httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/BasicHttpCache.java:334]: My suggestion was a general recommendation and aimed more at reducing log noise when evicting entries.

File: httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/BasicHttpAsyncCache.java [hunk 2]
```diff
@@ -456,57 +457,122 @@ public Cancellable storeReusing(
...
+    public Cancellable evictInvalidated(
             final HttpHost host, final HttpRequest request, final HttpResponse response, final FutureCallback<Boolean> callback) {
         if (LOG.isDebugEnabled()) {
             LOG.debug("Flush cache entries invalidated by exchange: {}; {} -> {}", host, new RequestLine(request), new StatusLine(response));
         }
-        if (!Method.isSafe(request.getMethod())) {
-            return cacheInvalidator.flushCacheEntriesInvalidatedByExchange(host, request, response, cacheKeyGenerator, storage, callback);
+        final int status = response.getCode();
+        if (status >= HttpStatus.SC_SUCCESS && status < HttpStatus.SC_CLIENT_ERROR &&
+                !Method.isSafe(request.getMethod())) {
```
[the remark this answers is not in the document]
Comment (ok2c) [reply httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/BasicHttpAsyncCache.java:562]: @arturobernalg Should not it be the other way around, cheapest conditions first, most expensive last?

[the remark this answers is not in the document]
Comment (arturobernalg) [reply httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/BasicHttpAsyncCache.java:562]: @ok2c  Apologies for the confusion caused by my previous comment. Upon further reflection, it's clear that the order of conditions in the if-statement is already well-optimized for short-circuiting.

**Pull request merged**